File: go/webhooks.go
package clarian

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

// Webhook replay window (matches server dispatch).
const webhookTolerance = 5 * time.Minute

// Header names on every Clarian webhook delivery.
const (
	HeaderSignature      = "X-Clarian-Signature"
	HeaderTimestamp      = "X-Clarian-Timestamp"
	HeaderEvent          = "X-Clarian-Event"
	HeaderDeliveryID     = "X-Clarian-Delivery-Id"
	HeaderIdempotencyKey = "X-Clarian-Idempotency-Key"
	HeaderAttempt        = "X-Clarian-Attempt"
)

// SignPayload returns the hex HMAC-SHA256 of timestamp+"."+payload using secret.
// Use with VerifyWebhook in local handler tests (inverse of the verify step).
func SignPayload(secret, timestamp string, payload []byte) string {
	mac := hmac.New(sha256.New, []byte(secret))
	_, _ = mac.Write([]byte(timestamp + "." + string(payload)))
	return hex.EncodeToString(mac.Sum(nil))
}

// VerifyWebhook checks HMAC-SHA256 signature and timestamp freshness, then
// parses the envelope. Signature = hex(HMAC-SHA256(secret, timestamp+"."+rawBody))
// where secret is the full whsec_… string (prefix included). Fail-closed on
// missing secret, signature, or timestamp.
func VerifyWebhook(payload []byte, headers http.Header, secret string) (*Event, error) {
	sig := headers.Get(HeaderSignature)
	tsRaw := headers.Get(HeaderTimestamp)
	if secret == "" || sig == "" || tsRaw == "" {
		return nil, fmt.Errorf("clarian: missing webhook signature, timestamp, or secret")
	}

	ts, err := time.Parse(time.RFC3339Nano, tsRaw)
	if err != nil {
		return nil, fmt.Errorf("clarian: invalid webhook timestamp: %w", err)
	}
	diff := time.Since(ts)
	if diff < 0 {
		diff = -diff
	}
	if diff > webhookTolerance {
		return nil, fmt.Errorf("clarian: webhook timestamp outside tolerance")
	}

	mac := hmac.New(sha256.New, []byte(secret))
	_, _ = mac.Write([]byte(tsRaw + "." + string(payload)))
	expected := hex.EncodeToString(mac.Sum(nil))
	if !hmac.Equal([]byte(expected), []byte(sig)) {
		return nil, fmt.Errorf("clarian: invalid webhook signature")
	}

	var envelope struct {
		ID          string          `json:"id"`
		Type        string          `json:"type"`
		CreatedAt   string          `json:"created_at"`
		Environment string          `json:"environment"`
		Data        json.RawMessage `json:"data"`
	}
	if err := json.Unmarshal(payload, &envelope); err != nil {
		return nil, fmt.Errorf("clarian: decode webhook: %w", err)
	}

	evt := &Event{
		ID:          envelope.ID,
		Type:        envelope.Type,
		CreatedAt:   envelope.CreatedAt,
		Environment: envelope.Environment,
		RawData:     envelope.Data,
	}
	if len(envelope.Data) > 0 {
		if err := json.Unmarshal(envelope.Data, &evt.Data); err != nil {
			return nil, fmt.Errorf("clarian: decode webhook data: %w", err)
		}
	}
	return evt, nil
}

File: php/src/Webhooks.php
<?php

declare(strict_types=1);

namespace Clarian;

/**
 * Webhook signature helpers.
 *
 * Signature is hex(HMAC-SHA256(secret, timestamp + "." + body)) using the full
 * whsec_… secret (prefix included; do not strip or base64-decode it).
 */
final class Webhooks
{
    public const HEADER_SIGNATURE = 'X-Clarian-Signature';
    public const HEADER_TIMESTAMP = 'X-Clarian-Timestamp';
    public const HEADER_EVENT = 'X-Clarian-Event';
    public const HEADER_DELIVERY_ID = 'X-Clarian-Delivery-Id';
    public const HEADER_IDEMPOTENCY_KEY = 'X-Clarian-Idempotency-Key';
    public const HEADER_ATTEMPT = 'X-Clarian-Attempt';

    public const DEFAULT_TOLERANCE_SECONDS = 300;

    private function __construct()
    {
    }

    /**
     * Return hex HMAC-SHA256 of "{timestamp}.{body}" using secret.
     */
    public static function signPayload(string $secret, string $timestamp, string $payload): string
    {
        return hash_hmac('sha256', $timestamp . '.' . $payload, $secret);
    }

    /**
     * Verify HMAC-SHA256 signature and timestamp freshness.
     *
     * Returns false on missing inputs, bad signature, or stale/future timestamp.
     */
    public static function verifySignature(
        string $payload,
        string $timestamp,
        string $signature,
        string $secret,
        int $toleranceSeconds = self::DEFAULT_TOLERANCE_SECONDS,
    ): bool {
        if ($secret === '' || $signature === '' || $timestamp === '') {
            return false;
        }

        try {
            $eventTs = self::parseTimestampEpoch($timestamp);
        } catch (\InvalidArgumentException) {
            return false;
        }

        $age = abs(time() - $eventTs);
        if ($age > $toleranceSeconds) {
            return false;
        }

        $expected = self::signPayload($secret, $timestamp, $payload);

        return hash_equals($expected, $signature);
    }

    /**
     * Read Clarian delivery headers (case-insensitive).
     *
     * Keys match the TypeScript SDK: signature, timestamp, event, deliveryId,
     * idempotencyKey, attempt. Missing values become empty strings.
     *
     * @param array<string, mixed> $headers
     * @return array{
     *   signature: string,
     *   timestamp: string,
     *   event: string,
     *   deliveryId: string,
     *   idempotencyKey: string,
     *   attempt: string
     * }
     */
    public static function extractHeaders(array $headers): array
    {
        $lower = [];
        foreach ($headers as $key => $value) {
            $lower[strtolower((string) $key)] = self::firstValue($value);
        }

        return [
            'signature' => $lower[strtolower(self::HEADER_SIGNATURE)] ?? '',
            'timestamp' => $lower[strtolower(self::HEADER_TIMESTAMP)] ?? '',
            'event' => $lower[strtolower(self::HEADER_EVENT)] ?? '',
            'deliveryId' => $lower[strtolower(self::HEADER_DELIVERY_ID)] ?? '',
            'idempotencyKey' => $lower[strtolower(self::HEADER_IDEMPOTENCY_KEY)] ?? '',
            'attempt' => $lower[strtolower(self::HEADER_ATTEMPT)] ?? '',
        ];
    }

    private static function firstValue(mixed $value): string
    {
        if ($value === null) {
            return '';
        }
        if (is_array($value)) {
            if ($value === []) {
                return '';
            }
            $first = reset($value);

            return $first === null ? '' : (string) $first;
        }

        return (string) $value;
    }

    private static function parseTimestampEpoch(string $timestamp): int
    {
        // Unix epoch (integer or float string)
        if (is_numeric($timestamp)) {
            return (int) $timestamp;
        }

        // ISO 8601 / RFC3339 (with or without fractional seconds, Z or offset)
        $normalized = str_replace('Z', '+00:00', $timestamp);
        $dt = \DateTimeImmutable::createFromFormat(
            \DateTimeInterface::RFC3339_EXTENDED,
            $normalized,
        );
        if ($dt === false) {
            $dt = \DateTimeImmutable::createFromFormat(
                \DateTimeInterface::RFC3339,
                $normalized,
            );
        }
        if ($dt === false) {
            $dt = date_create_immutable($timestamp);
        }
        if ($dt === false) {
            throw new \InvalidArgumentException('invalid webhook timestamp: ' . $timestamp);
        }

        return $dt->getTimestamp();
    }
}
